go.mod

```diff
@@ -1,6 +1,6 @@
 module sigs.k8s.io/kubebuilder/v3
 
-go 1.17
+go 1.18
```

Comment:
Will this require all importers of `kubebuilder/v3` to also upgrade to 1.18?

If so, do we want to force that migration so soon after 1.18 was released?

I wonder if we can make the scaffolding support 1.18 without bumping the kubebuilder module itself to 1.18?

Reply:
1. Who's importing `kubebuilder/v3`? Also, they don't have to import `master` version - they can pin the version to the previous tag.
2. Not merging this will block users that build the binary locally (i.e. brew users)

Reply:
1. Anyone who has written a plugin. I know of at least projects in the operator-framework that import it. Not sure if there are others. If it was _just_ the operator-framework projects, I'd concede this isn't that big of a deal.
2. You're saying someone running Go 1.18 cannot build a Go module that declares 1.17 compatibility? That doesn't seem right since Go 1 has a back-compat guarantee. Or is it something else?

Reply:
a. When I run Kubebuilder that was built using 1.17 from my environment(that has 1.18) it throws some errors
b. When installing Kubebuilder using brew, it's locally compiles the binaries using go 1.18 - which leads to a failure

Reply:
I've tested it with `go.mod` specifying version 1.17 and it seems to work.
That being said, I don't think it makes sense to generate files for go 1.18 but not to require 1.18 for Kubebuilder:
That basically means we will not bump the version to go 1.18, but only "support" 1.18 - which will prevent us from using 1.18 features in the future.

If an upstream user wants to use go 1.17 but is not willing to upgrade, he can always specify a different version.

Reply:
@AlmogBaku, 
The concern here is to know for now who imports the Kubebuilder as LIB. Do we know what are the impacts that they can face? ( e.g SDK uses Kubebuilder as a lib, will SDK still be able to use it after the changes or to still importing KB will be required also upgrade the SDK project? ) 

Kubebuilder CLI users run the binary so for them only matters the go version in the scaffolds produced.

Reply:
by defining the `go.mod` to 1.18 we'll require upstream users to use 1.18, and on the other hand- we'll benefit from 1.18 features.

Users (such as the Operator-SDK) that haven't upgraded to 1.18 will have to either upgrade or pin the version to v3.3.0.

Reply:
As a side note- IMO, not upgrading go will make this PR **totally redundant**.

To solve compatibility issues, we just need to upgrade a few dependencies, and that's it. But then we're not upgrading anything.

Reply:
I think there's some detail lacking in this thread.

> a. When I run Kubebuilder that was built using 1.17 from my environment(that has 1.18) it throws some errors
> b. When installing Kubebuilder using brew, it's locally compiles the binaries using go 1.18 - which leads to a failure

What errors, what failure?

> By defining the go.mod to 1.18 we'll require upstream consumers to use 1.18, and on the other hand- we'll benefit from 1.18 features.

Are we introducing changes in the Kubebuilder CLI codebase that require Go 1.18? If so, can we call those out? If not, it seems better to wait until we need those 1.18 features before we force library consumers to upgrade their go version.

I'm not saying we definitely shouldn't bump the kubebuilder go.mod go version to 1.18, I'm just asking for the specific motivation for that. Kubebuilder's go dependencies can be confusing because we're talking about at least 2 different things:
1. The version of Go used to compile Kubebuilder
2. The version of Go that the code _scaffolded_ by Kubebuilder uses.

As far as I know, these are completely disjoint from each other (after all, I could write a python program that scaffolds Go code the exact same way that Kubebuilder does). Is there something we've done that couples these versions somehow?

I think the motivation for 2 is extremely obvious (we want users of the Kubebuilder CLI to be able to build projects with the latest and greatest Go version). But for 1, it seems like we're saying we _need_ to bump the Kubebuilder go.mod in order to support scaffolding 1.18 code. That's where I'm not following.

Other than my question about whether we're ready to force kubebuilder importers to use 1.18, I don't have any other qualms with bumping the kubebuilder go.mod.

I'd just like to see an explanation for why we're bumping the kubebuilder go.mod. If it's reasonable or pragmatic, great let's do it! I just think its important to have a complete understanding and consensus for why we we're doing it.

Reply:
The reason is to prepare the ground for future contributions that require go 1.18.

Not doing so now will just postpone the decision.
Imagine a contributor that wants to add a simple fix using Generics, because we haven't taken the call yet - we'll have to ask him to rewrite the solution without Generics.

Reply:
I guess I'd imagine that that PR would be their generics change, plus the one liner in the go.mod file. Is it more than that?

Reply:
I could see us saying: Oh, we need to bump to 1.18 for that. Let's block your PR on this other 1.18 bump PR. But at least then there's a specific motivation to bump and we can debate the merits of requiring 1.18 vs staying at 1.17.

We could even debate your point that _right now_ if we separated this PR out into a) bump KB to 1.18 and b) bump the scaffolding to 1.18.

Reply:
+1 for @joelanford comments
I do not think that using generics in the Kubebuilder API is something that we will do soon at all.
Anyway, we also identified that we ought to bump the k8s api with and update the CLI/API to use go 1.18 seems that we also need to update the controller tools with go 1.18 since config-gen is using this project as a lib.